api/rbac/rbac.go

```diff
@@ -0,0 +1,191 @@
+// everest
+// Copyright (C) 2023 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package rbac provides RBAC middleware utilies for the Everest API server.
+package rbac
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"slices"
+	"strings"
+
+	"github.com/casbin/casbin/v2"
+	"github.com/casbin/casbin/v2/model"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/labstack/echo/v4"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	configmapadapter "github.com/percona/everest/api/rbac/configmap-adapter"
+	"github.com/percona/everest/data"
+	"github.com/percona/everest/pkg/common"
+	"github.com/percona/everest/pkg/kubernetes"
+	"github.com/percona/everest/pkg/kubernetes/informer"
+	"github.com/percona/everest/pkg/session"
+)
+
+// Setup a new informer that watches our RBAC ConfigMap.
+// This informer reloads the policy whenever the ConfigMap is updated.
+func refreshEnforcerInBackground(
+	ctx context.Context,
+	kubeClient *kubernetes.Kubernetes,
+	enforcer *casbin.Enforcer,
+	l *zap.SugaredLogger,
+) error {
+	inf, err := informer.New(
+		informer.WithConfig(kubeClient.Config()),
+		informer.WithLogger(l),
+		informer.Watches(&corev1.ConfigMap{}, kubeClient.Namespace()),
+	)
+	inf.OnUpdate(func(_, newObj interface{}) {
+		cm, ok := newObj.(*corev1.ConfigMap)
+		if !ok || cm.GetName() != common.EverestRBACConfigMapName {
+			return
+		}
+		if err := enforcer.LoadPolicy(); err != nil {
+			l.Error("failed to load policy", zap.Error(err))
+		}
+	})
+	if inf.Start(ctx, &corev1.ConfigMap{}) != nil {
+		return errors.Join(err, errors.New("failed to watch RBAC ConfigMap"))
+	}
+	return nil
+}
```

Comment:
This is such an important piece that was missing from the PoC, thanks for adding this hot reload :clap: